cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql

```diff
@@ -15,63 +15,113 @@
 import cpp
 import semmle.code.cpp.security.CommandExecution
 import semmle.code.cpp.security.Security
-import DataFlow::PathGraph
 import semmle.code.cpp.ir.dataflow.TaintTracking
+import semmle.code.cpp.ir.dataflow.TaintTracking2
 import semmle.code.cpp.ir.IR
 import semmle.code.cpp.security.FlowSources
+import semmle.code.cpp.models.implementations.Strcat
 
 Expr sinkAsArgumentIndirection(DataFlow::Node sink) {
-  result = sink.asOperand()
-      .(SideEffectOperand)
-      .getAddressOperand()
-      .getAnyDef()
-      .getUnconvertedResultExpression()
+  result =
+    sink.asOperand()
+        .(SideEffectOperand)
+        .getAddressOperand()
+        .getAnyDef()
+        .getUnconvertedResultExpression()
 }
 
-class ExecTaintConfiguration extends TaintTracking::Configuration {
-  ExecTaintConfiguration() { this = "ExecTaintConfiguration" }
+predicate interestingConcatenation(DataFlow::Node fst, DataFlow::Node snd) {
+  exists(FormattingFunctionCall call, int index, FormatLiteral literal |
+    sinkAsArgumentIndirection(fst) = call.getConversionArgument(index) and
+    snd.asDefiningArgument() = call.getOutputArgument(false) and
+    literal = call.getFormat() and
+    not literal.getConvSpecOffset(index) = 0 and
+    (
+      literal.getConversionType(index) instanceof CharPointerType
+      or
+      literal.getConversionType(index).(PointerType).getBaseType() instanceof Wchar_t
+    )
+  )
+  or
+  // strcat and friends
+  exists(StrcatFunction strcatFunc, CallInstruction call, ReadSideEffectInstruction rse |
+    call.getStaticCallTarget() = strcatFunc and
+    rse.getArgumentDef() = call.getArgument(strcatFunc.getParamSrc()) and
+    fst.asOperand() = rse.getSideEffectOperand() and
+    snd.asInstruction().(WriteSideEffectInstruction).getDestinationAddress() =
+        call.getArgument(strcatFunc.getParamDest())
+  )
+  or
+  exists(CallInstruction call, Operator op, ReadSideEffectInstruction rse |
+    call.getStaticCallTarget() = op and
+    op.hasQualifiedName("std", "operator+") and
+    op.getType().(UserType).hasQualifiedName("std", "basic_string") and
+    call.getArgument(1) = rse.getArgumentOperand().getAnyDef() and // left operand
+    fst.asOperand() = rse.getSideEffectOperand() and
+    call =
+      snd.asInstruction()
+  )
+}
+
+// TODO: maybe we can drop this?
+class TaintToConcatenationConfiguration extends TaintTracking::Configuration {
```

Comment:
If I understand correctly, this strategy ensures that taint passes through at least one 'interesting' concatenation.  Two details spring to mind:
 - if there are no concatenations or string processing at all, just an `exec` of the tainted string as it is, currently I think this will not be reported.  Strictly speaking the tainted string *is* at the start of the `exec` string, but is this what we want to do?
 - we no longer treat `printf("%i", taint)` as a barrier, merely as a non-interesting concatenation, which I think will fail to rule out some multi-step cases that we'd have ruled out when it was a barrier.

Reply:
>  Strictly speaking the tainted string is at the start of the exec string, but is this what we want to do?

My assumption was that it would be a deliberate decision (like `codeql databse create -c`).

> we no longer treat printf("%i", taint) as a barrier

Good point. Maybe a barrier on any value with an integral type.

Reply:
Added